src/mongocxx/options/index.hpp

```diff
@@ -0,0 +1,365 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongocxx/config/prelude.hpp>
+
+#include <bsoncxx/document/view.hpp>
+#include <bsoncxx/stdx/optional.hpp>
+#include <mongocxx/stdx.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+
+class collection;
+
+namespace options {
+
+///
+/// Class representing the optional arguments to a MongoDB create index operation.
+///
+/// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+///
+class MONGOCXX_API index {
+   public:
+    ///
+    /// Base class representing the optional storage engine options for indexes.
+    ///
+    class base_storage_options {
+       private:
+        friend collection;
+        MONGOCXX_PRIVATE virtual const int type() const = 0;
+    };
+
+    ///
+    /// Class representing the optional WiredTiger storage engine options for indexes.
+    ///
+    class wiredtiger_storage_options : public base_storage_options {
+       public:
+        ///
+        /// Set the WiredTiger configuration string.
+        ///
+        /// @param config_str
+        ///   The WiredTiger configuration string.
+        ///
+        void config_str(std::string config_str);
+
+        ///
+        /// The current config_str setting.
+        ///
+        /// @return The current config_str.
+        ///
+        const stdx::optional<std::string>& config_str() const;
+
+       private:
+        friend collection;
+        MONGOCXX_PRIVATE const int type() const override;
+        stdx::optional<std::string> _config_str;
+    };
+
+    index();
+
+    ///
+    /// Whether or not to build the index in the background so that building
+    /// the index does not block other database activities.
+    ///
+    /// @param background
+    ///   Whether or not to build the index in the background.
+    ///
+    /// @see https://docs.mongodb.org/manual/tutorial/build-indexes-in-the-background/
+    ///
+    void background(bool background);
+
+    ///
+    /// The current background setting.
+    ///
+    /// @return The current background.
+    ///
+    const stdx::optional<bool>& background() const;
+
+    ///
+    /// Whether or not to create a unique index so that the collection will not
+    /// accept insertion of documents where the index key or keys match an
+    /// existing value in the index.
```

Comment:
I did mean "value in the index" here. I thought that for unique indexes, the indexes would not insert values already in the index. Do you think the other way is more clear?

(also, sorry, this line will be / was changed by clang format)